packages/tiled/lib/src/common/property.dart

```diff
@@ -12,35 +12,213 @@ part of tiled;
+    final type = parser.getPropertyType('type', defaults: PropertyType.string);
+
+    switch (type) {
+      case PropertyType.object:
+        return ObjectProperty(
+          name: name,
+          value: parser.getInt('value', defaults: 0),
+        );
+
+      case PropertyType.color:
+        return ColorProperty(
+          name: name,
+          value: parser.getColor('value', defaults: const Color(0x00000000)),
+          hexValue: parser.getString('value', defaults: '#00000000'),
+        );
+
+      case PropertyType.bool:
+        return BoolProperty(
+          name: name,
+          value: parser.getBool('value', defaults: false),
+        );
+
+      case PropertyType.float:
+        return FloatProperty(
+          name: name,
+          value: parser.getDouble('value', defaults: 0),
+        );
+
+      case PropertyType.int:
+        return IntProperty(
+          name: name,
+          value: parser.getInt('value', defaults: 0),
+        );
+
+      case PropertyType.file:
+        return FileProperty(
+          name: name,
+          value: parser.getString('value', defaults: '.'),
+        );
+
+      case PropertyType.string:
+        final value = parser.formatSpecificParsing((json) {
+          return json.getString('value', defaults: '');
+        }, (xml) {
+          final attrString = parser.getStringOrNull('value');
+          if (attrString != null) {
+            return attrString;
+          } else {
+            // In tmx files, multi-line text property values can be stored
+            // inside the <property> node itself instead of in the 'value'
+            // attribute
+            return xml.element.innerText;
+          }
+        });
+
+        return StringProperty(
+          name: name,
+          value: value,
         );
+    }
+  }
+}
+
+/// A wrapper for a Tiled property set
+///
+/// Accessing an int value
+/// ```dart
+/// properties.get<int>('foo') == 3
+/// ```
+///
+/// Accessing an int property:
+/// ```dart
+/// properties.getProp<IntProperty>('foo') ==
+///     IntProperty(name: 'foo', value: 3);
+/// ```
+class CustomProperties extends Iterable<Property<Object>> {
+  static const empty = CustomProperties({});
+
+  /// The properties, indexed by name
+  final Map<String, Property<Object>> byName;
+
+  const CustomProperties(this.byName);
+
+  /// Get a property value by its name.
+  ///
+  /// [T] must be the type of the value, which depends on the property type.
+  /// The following Tiled properties map to the follow Dart types:
+  ///  - int property -> int
+  ///  - float property -> double
+  ///  - boolean property -> bool
+  ///  - string property -> string
+  ///  - color property -> ui.Color
+  ///  - file property -> string (path)
+  ///  - object property -> int (ID)
+  T? getValue<T>(String name) {
+    return getProperty(name)?.value as T?;
+  }
+
+  /// Get a typed property by its name
+  T? getProperty<T extends Property<Object>>(String name) {
```

Comment:
```suggestion
  T? getProperty<T extends Property<T>>(String name) {
```

Reply:
`T` is the type of the property here, not the value, so this change would require that the Property has a value of Property recursively. We'd have to add a second generic to the method, `K`, which would be awkward:
`getProperty<IntProperty, int>('foo')`

Reply:
Ah, of course. Wouldn't one want to just want to specify the type of the value instead, and then the return type can be `Property<T>?`?

Reply:
Well the intent was to make it simple to get an `IntProperty` or ` ColorProperty`, otherwise you still have to cast it to one of those if you want the actual property type, which would be more relevant if the property has unique methods on it. Otherwise there is really no reason to be grabbing the property itself at all, you can just grab the value directly

Reply:
Hmm, what is the real point of having the property wrapper actually? Can't we just offer extensions on the value type instead? Like `toHex` on `Color` for example?

Reply:
We may not need it, but it's to entirely clear to me that every property can just be a simple name/value pair where the value doesn't need custom behavior, if so we should probably ditch the entire Property class

Reply:
For all the `value` based use cases your describing, that's what `getValue<T>` is for, right?

Reply:
So, are there any usecases for having this wrapper class? @ufrshubham @jtmcdole

Reply:
I'll also say that there is some value to me to making the class structure match the Tiled data structures for consistency with Tiled itself. Also, we really do want to know if a property is a file string or a text string, etc